pkg/pillar/containerd/containerd.go

```diff
@@ -229,22 +178,48 @@ func CtrPrepare(containerPath string, ociFilename string) error {
 	log.Infof("mountPoints %+v execpath %+v cmd %+v workdir %+v env %+v",
 		mountpoints, execpath, cmd, workdir, unProcessedEnv)
 
-	if err != nil {
-		return fmt.Errorf("ctrPrepare: unable to get image: %v config: %v", ctrdImage.Name(), err)
-	}
-	// containerd.NewImageWithPlatform(client, i, platforms.Only(platform))
+	// unpack the rootfs Image if needed
 	unpacked, err := ctrdImage.IsUnpacked(ctrdCtx, defaultSnapshotter)
 	if err != nil {
-		return fmt.Errorf("ctrPrepare: unable to get image metadata: %v config: %v", ctrdImage.Name(), err)
+		return fmt.Errorf("snapshotPrepare: unable to get image metadata: %v config: %v", ctrdImage.Name(), err)
 	}
 	if !unpacked {
 		if err := ctrdImage.Unpack(ctrdCtx, defaultSnapshotter); err != nil {
-			return fmt.Errorf("ctrPrepare: unable to unpack image: %v config: %v", ctrdImage.Name(), err)
+			return fmt.Errorf("snapshotPrepare: unable to unpack image: %v config: %v", ctrdImage.Name(), err)
 		}
 	}
-	if err = ctrCreate(filepath.Base(containerPath), ctrdImage); err != nil {
-		return fmt.Errorf("ctrPrepare: failed to create container %s, error: %v", containerPath, err.Error())
+
+	// use rootfs unpacked image to create a writable snapshot with default snapshotter
+	diffIDs, err := ctrdImage.RootFS(ctrdCtx)
+	if err != nil {
+		log.Errorf("Could not load rootfs of image: %v. %v", ctrdImage.Name(), err)
+		return fmt.Errorf("snapshotPrepare: Could not load rootfs of image: %v. %v", ctrdImage.Name(), err)
+	}
+
+	snapshotter := CtrdClient.SnapshotService(defaultSnapshotter)
+	parent := identity.ChainID(diffIDs).String()
+	snapshot := getSnapshotName(containerID)
+	labels := map[string]string{"containerd.io/gc.root": time.Now().UTC().Format(time.RFC3339)}
```

Comment:
While this will work, the canonical way to do this is to add a `containerd.io/gc.ref.snapshot.overlayfs=sha256:<sha-of-parent>` to whichever blob is holding the config in the content store.

This is fine, as long as we know that we are taking some responsibility here for removing it, otherwise it, and all of its children, will linger.

Reply:
Actually, this _is_ unnecessary. We are doing `Unpack()` earlier (see line 187), which adds he correct label to the config blob. See [here](https://github.com/containerd/containerd/blob/master/image.go#L354) in the source code.

So we can get out of the business of adding these labels entirely.

Reply:
Somehow it doesn't @deitch -- while agree with you it should -- it cost me 2 hours of my life to bang my head at the wall seeing that GC always kicks in and deletes snapshot I just created. Then I decided to throw in the towel and simply pin it explicitly. Since we manage snapshots very explicitly in volumemgr -- I'm not worried about any kind of leakage.

On the flip side -- if you can help figure out this mystery -- that would be very much appreciated  @deitch

Reply:
I think you may be right. Unless you create a container to hold onto it, there is no way other than a label like this.